### PR TITLE
[FLOC-3271] Acceptance test for Flocker for storage profiles.

### PR DESCRIPTION
**Pull request by sarum90**
This is the codified acceptance test for Flocker storage profiles basic support on EBS

It verifies that if a volume is created with the specific metadata that forces a profile for the volume, then the volume that is created has the correct EBS type for the given profile.

In the interest of keeping the acceptance tests lean, I've only added a test for gold rather than also testing for silver and bronze. I'm open to feedback on that decision. I figured this was enough to show the plumbing was there even if we aren't exhaustively testing every length of pipe at this level.

In order to validate the volume's performance characteristics from the acceptance test, this adds a backdoor to get the backend configuration from the process that sets up the acceptance test rig.  It then uses those credentials to verify the type of the volume created.

[Review on Reviewable](https://reviewable.io/reviews/clusterhq/flocker/2094)

### CODE REVIEW

**Comment by sarum90** [comment]
I ran the tests once before adding the SkipTest (http://build.clusterhq.com/boxes-flocker?branch=flocker-storage-profiles-acceptance-FLOC-3271) if we're lucky they'll fail for all the right reasons, specifically that the "standard" volume type is not "io1" the type expected for gold (they did on my local runs). Then I'll re-kick them and we should see this new test being skipped. I'll take care of that in address_and_merge. I don't think that should block review though. If you have a different opinion let me know and I'll just sit on this until that testing is complete.

File: flocker/acceptance/testtools.py [hunk 1]
```diff
@@ -789,6 +839,15 @@ def decorator(test_method):
         """
         def call_test_method_with_cluster(cluster, test_case, args, kwargs):
             kwargs['cluster'] = cluster
+            if required_backend:
+                backend_type = get_dataset_backend(test_case)
+                if backend_type != required_backend:
+                    raise SkipTest(
+                        'This test requires backend type {} but is being run '
+                        'on {}.'.format(backend_type.name,
```
Comment (sarum90) [inline flocker/acceptance/testtools.py]: These are backwards: http://build.clusterhq.com/builders/flocker%2Facceptance%2Frackspace%2Fcentos-7%2Frackspace/builds/3001/steps/trial/logs/stdio

I'll fix that when I address other review comments.

File: flocker/acceptance/testtools.py [hunk 2]
```diff
@@ -866,10 +928,12 @@ def create_dataset(test_case, cluster, maximum_size=None, dataset_id=None):
         maximum_size = get_default_volume_size()
     if dataset_id is None:
         dataset_id = uuid4()
-
+    if metadata is None:
+        metadata = {}
+    metadata[u"name"] = metadata.get(u"name", u"my_volume")
```
Comment (itamarst) [inline flocker/acceptance/testtools.py]: Mutating inputs isn't a great idea, tests might accidently pass-in something that then gets mutated. Make a copy and mutate that.

File: flocker/provision/_common.py [hunk 3]
```diff
@@ -71,6 +71,8 @@ class Cluster(PRecord):
     :ivar FilePath certificates_path: Directory where certificates can be
         found; specifically the directory used by ``Certificates``.
     :ivar Certificates certificates: Certificates to for the cluster.
+    :ivar unicode dataset_backend_config_file: File with the backend
```
Comment (itamarst) [inline flocker/provision/_common.py]: FilePath instead?

**Comment by itamarst** [comment]
Thinking out loud:

1. Access to backend is useful, I guess. Sometimes the effects we care about are in external systems.
2. I'm not sure this is best way to do it, but then I don't have any better ideas.
3. This will need to support other backends.

But... I guess this is interim step and long term plan is to expose profile info in REST API, and then we don't need to talk to the backend and can test that way?

**Comment by itamarst** [comment]
So... address and merge. Still not sure if this is a good idea or a bad idea, but it's definitely an interesting idea! :grinning:

File: flocker/acceptance/testtools.py (hunk 1 above)
Comment (sarum90) [inline flocker/acceptance/testtools.py]: Fixed

File: flocker/acceptance/testtools.py (hunk 2 above)
Comment (sarum90) [inline flocker/acceptance/testtools.py]: Thanks for catching that! Fixed

File: flocker/provision/_common.py (hunk 3 above)
Comment (sarum90) [inline flocker/provision/_common.py]: Done

**Comment by sarum90** [comment]
+1 on it maybe not being the best idea, still seemed better than not having codified acceptance criteria.

Re:supporting other backends. If and when we need that should be simple to code. We would have to rework a bit of the node agent so that the "instantiate IBlockDeviceAPI" code could be re-used. Figured this was good enough for now for the current acceptance criteria. On the other hand... if we get to the point where we no longer need this... then we can just remove it.

Didn't just work on Buildbot though... going to look into that.

**Comment by sarum90** [comment]
Enabled the test, saw it fail on Buildbot for the right reason. Disabled the test, got a clean buildbot run.

Merging.

**Pull request merged**